### PR TITLE
Add a GitHub Action to automatically rebuild files

### LINKED ISSUE #1224: Use github actions to generate _config.yml
currently we rely on an obo maintainer to manually run the makefile to compile any changes to ontology/, then commit push results

We should just use github actions for this. It's easy to make a custom workflow that will run the make step.

Note: if we wanted to move away from hosting on gh pages, e.g. static site served off of S3, there are actions for this:

https://github.com/marketplace/actions/build-jekyll

### PR DESCRIPTION
**Pull request by cthoyt**
Closes #1224

This PR adds a new GitHub Action configuration, `build.yml` that does the following things:

1. Installs ODK container
2. Installs the Python requirements from this repository's `requirements.txt` 
3. Runs `make`
4. Commits/pushes the result back to the `master` branch

It automatically runs any time a markdown file is changed in the `ontology/` folder on the master branch. This means that there's no longer a need for a second set of discussions, PRs, etc. every time a metadata file is updated. In theory, the build process is automated and does not itself need review. This also has the added benefit that potential contributors can see their updates incorporated much faster.

This action can also be run on demand, in case some of the tooling changes and you don't want to wait for the next change to a markdown file.

You can see the workflow in action at https://github.com/cthoyt/OBOFoundry.github.io/runs/4226590920?check_suite_focus=true and see the resulting commit it made in https://github.com/cthoyt/OBOFoundry.github.io/commit/663461070fb5889be058863069530e05251d3419. Note it's a pretty boring commit due to the fact that no markdown files were changed. Unfortunately, the RDF dump is not deterministic so this is actually a completely spurious diff. Either way, it demonstrates that everything runs properly.

### CODE REVIEW

**Comment by cthoyt** [comment]
Question: does https://github.com/OBOFoundry/OBOFoundry.github.io#generation-of-downstream-artefacts need to be updated? I'm not sure if the jekyll step is necessary or not. It's currently not covered by this PR

**Comment by matentzn** [comment]
Awesome, thanks! I think this may be the right move, but so far we have been more conservative as we (used to) frequently spot errors to metadata in particular during the overall review of ontologies.yml and the various context files. I am right this moment on the fence whether we should commit to master. I would prefer a pull request, but only just. if @cmungall or @jamesaoverton or others on the technical team have different opinions I am also ok to risk it. You may want to use odklite instead of odkfull which is a much smaller container.

**Comment by cthoyt** [comment]
@matentzn can you better describe what kind of errors you could catch by looking at the output of the build that wouldn't be possible to find by looking at the diff / using the CI tests before merging the actual PR that changes the underlying metadata files?

The last 2 PRs I found #1652 and #1349 didn't have any comments related to the regeneration of the files. In #1652, there were actually additional files that got edited - I think it would be good to specifically avoid conflating building files with actual edits

**Comment by matentzn** [comment]
Ok, seems like in the associated [issue](#1224), we agreed on risking it.. Lets roll with it for a while and retreat when there is an issue. Thanks!

For the record, i am thinking of things like when our context is generated, it generates for example a curie map. This curie map has been wrong in the past when people did record the preferred prefix correctly. 

Anyways triffle. What was I thinking. BRAVENESS! AUTOMATION!

**Review by matentzn — APPROVED** [review]
Nice one!

**Comment by jamesaoverton** [comment]
Much appreciated. This looks good, but I have one question. In our GitHub repo settings we have `master` as a protected branch, requiring PR approval. Does the GitHub Action ignore the branch protection? I think ignoring the branch protection is the behaviour we want, but I don't know what will actually happen.

**Review by jamesaoverton — APPROVED** [review]

**Comment by jamesaoverton** [comment]
Ok, let's see how this goes...

**Comment by matentzn** [comment]
currently not, see #1662

**Comment by cthoyt** [comment]
Screen Shot 2021-11-17 at 19 05 58

I realized in PyKEEN we have it set up so the bot that makes automatic pushes has explicit access rights via its membership in a team. So we can do the same for the GitHub Actions bot in this rep

**Comment by matentzn** [comment]
Will this work with 

- [X] Include administrators?

image

**Comment by jamesaoverton** [comment]
Ok. I implemented some of #1404, so now there are just six Admins. Then I removed the "Include administrators" restriction. Maybe it will work now?

There's also the "Allow specified actors to bypass pull request requirements" option. But how would I specify GitHub Actions is allowed to bypass?

**Comment by jamesaoverton** [comment]
Failed again: https://github.com/OBOFoundry/OBOFoundry.github.io/runs/4253718821?check_suite_focus=true

**Comment by matentzn** [comment]
@cthoyt do we have to:

- Check the option for restrict who can push to matching branches
- Add github actions bot to admin team?

**Pull request merged**
